kolibri/content/api.py

```diff
@@ -111,30 +122,48 @@ def intersection(queries):
         def union(queries):
             return reduce(lambda x, y: x | y, queries)
 
+        # all words with punctuation removed
+        all_words = [w for w in re.split('[?.,!";: ]', value) if w]
+        # words in all_words that are not stopwords
+        critical_words = [w for w in all_words if w not in stopwords_set]
+        # stopwords from the json file
+        stop_words = list(set(all_words).difference(critical_words))
         # queries ordered by relevance priority
-        words = value.split()
         all_queries = [
             # exact match in title
             Q(title__icontains=value),
             # all words in title
-            intersection([Q(title__icontains=w) for w in words]),
+            intersection([Q(title__icontains=w) for w in all_words]),
+            # all critical words in title
+            intersection([Q(title__icontains=w) for w in critical_words]),
             # exact match in description
             Q(description__icontains=value),
             # all words in description
-            intersection([Q(description__icontains=w) for w in words]),
+            intersection([Q(description__icontains=w) for w in all_words]),
+            # all critical words in description
+            intersection([Q(description__icontains=w) for w in critical_words]),
         ]
-
-        # any word in title, sorted by length
-        for w in reversed(sorted(words, key=len)):
+        # any critical word in title, reverse-sorted by word length
+        for w in sorted(critical_words, key=len, reverse=True):
             all_queries.append(Q(title__icontains=w))
+        # any critical word in description, reverse-sorted by word length
+        for w in sorted(critical_words, key=len, reverse=True):
+            all_queries.append(Q(description__icontains=w))
+        # any stop word in title, reverse-sorted by word length
+        for w in sorted(stop_words, key=len, reverse=True):
+            all_queries.append(Q(title__icontains=w))
+        # any stop word in description, reverse-sorted by word length
+        for w in sorted(stop_words, key=len, reverse=True):
+            all_queries.append(Q(description__icontains=w))
```

Comment:
To shave a little time off these searches, maybe we can reduce the number of queries. Something like:

```python
# all words with punctuation removed
all_words = [w for w in re.split('[?.,!";: ]', value) if w]
# words in all_words that are not stopwords
critical_words = [w for w in all_words if w not in stopwords_set]
# queries ordered by relevance priority
all_queries = [
    # all words in title
    intersection([Q(title__icontains=w) for w in all_words]),
    # all critical words in title
    intersection([Q(title__icontains=w) for w in critical_words]),
    # all words in description
    intersection([Q(description__icontains=w) for w in all_words]),
    # all critical words in description
    intersection([Q(description__icontains=w) for w in critical_words]),
]
# any word in title, reverse-sorted by word length
for w in sorted(all_words, key=len, reverse=True):
    all_queries.append(Q(title__icontains=w))
# any word in description, reverse-sorted by word length
for w in sorted(all_words, key=len, reverse=True):
    all_queries.append(Q(description__icontains=w))
```